File: backend/etfpulse/api/schemas/execution.py
"""Pydantic shapes for the PR D.4.3 execution surface.

These are the API boundary — strings on the wire, mapped to D.1 SoDEX
IntEnums inside the route handler before the call into D.3's risk +
pipeline. The map lives in this module (not on the orchestrator) so
adding a new API enum value doesn't ripple into `pipeline/execution/`.
"""

from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from enum import StrEnum

from pydantic import BaseModel, Field, model_validator

from etfpulse.adapters.sodex.schemas import (
    OrderSide as SodexOrderSide,
)
from etfpulse.adapters.sodex.schemas import (
    OrderType as SodexOrderType,
)
from etfpulse.adapters.sodex.schemas import (
    PositionSide as SodexPositionSide,
)
from etfpulse.adapters.sodex.schemas import (
    TimeInForce as SodexTimeInForce,
)
from etfpulse.adapters.sodex.schemas import (
    TriggerType as SodexTriggerType,
)
from etfpulse.models.order import StopType, Venue

# ---------------------------------------------------------------------------
# Wire enums (lowercase strings) + map to SoDEX IntEnums
# ---------------------------------------------------------------------------


class ApiSide(StrEnum):
    BUY = "buy"
    SELL = "sell"


class ApiOrderType(StrEnum):
    LIMIT = "limit"
    MARKET = "market"


class ApiTimeInForce(StrEnum):
    GTC = "gtc"
    IOC = "ioc"
    FOK = "fok"
    GTX = "gtx"


class ApiPositionSide(StrEnum):
    BOTH = "both"
    LONG = "long"
    SHORT = "short"


class ApiTriggerType(StrEnum):
    MARK_PRICE = "mark_price"
    LAST_PRICE = "last_price"
    INDEX_PRICE = "index_price"


class ApiStopType(StrEnum):
    """Wire-side mirror of `models.order.StopType`. Values match the DB
    CHECK literals exactly (`'stop_loss'` / `'take_profit'`) so the
    schema can pass `Order.stop_type` round-trips without conversion."""

    STOP_LOSS = StopType.STOP_LOSS.value
    TAKE_PROFIT = StopType.TAKE_PROFIT.value


# Single-source-of-truth maps. Adding a new wire-enum value: add to the
# StrEnum above + add the mapping here. The risk gate downstream catches
# any value the gateway doesn't actually accept (e.g., FOK is on the
# enum but `_UNSUPPORTED_TIME_IN_FORCE` rejects it).
_API_TO_SODEX_SIDE: dict[str, int] = {
    ApiSide.BUY.value: SodexOrderSide.BUY.value,
    ApiSide.SELL.value: SodexOrderSide.SELL.value,
}
_API_TO_SODEX_TYPE: dict[str, int] = {
    ApiOrderType.LIMIT.value: SodexOrderType.LIMIT.value,
    ApiOrderType.MARKET.value: SodexOrderType.MARKET.value,
}
_API_TO_SODEX_TIF: dict[str, int] = {
    ApiTimeInForce.GTC.value: SodexTimeInForce.GTC.value,
    ApiTimeInForce.IOC.value: SodexTimeInForce.IOC.value,
    ApiTimeInForce.FOK.value: SodexTimeInForce.FOK.value,
    ApiTimeInForce.GTX.value: SodexTimeInForce.GTX.value,
}
_API_TO_SODEX_POSITION_SIDE: dict[str, int] = {
    ApiPositionSide.BOTH.value: SodexPositionSide.BOTH.value,
    ApiPositionSide.LONG.value: SodexPositionSide.LONG.value,
    ApiPositionSide.SHORT.value: SodexPositionSide.SHORT.value,
}
_API_TO_SODEX_TRIGGER_TYPE: dict[str, int] = {
    ApiTriggerType.MARK_PRICE.value: SodexTriggerType.MARK_PRICE.value,
    ApiTriggerType.LAST_PRICE.value: SodexTriggerType.LAST_PRICE.value,
    ApiTriggerType.INDEX_PRICE.value: SodexTriggerType.INDEX_PRICE.value,
}


def api_side_to_sodex(v: str) -> int:
    return _API_TO_SODEX_SIDE[v]


def api_order_type_to_sodex(v: str) -> int:
    return _API_TO_SODEX_TYPE[v]


def api_tif_to_sodex(v: str) -> int:
    return _API_TO_SODEX_TIF[v]


def api_position_side_to_sodex(v: str) -> int:
    return _API_TO_SODEX_POSITION_SIDE[v]


def api_trigger_type_to_sodex(v: str) -> int:
    return _API_TO_SODEX_TRIGGER_TYPE[v]


# ---------------------------------------------------------------------------
# Request models
# ---------------------------------------------------------------------------


class PrepareNewRequest(BaseModel):
    """`POST /api/execution/prepare` body.

    Mirrors `pipeline.execution.risk.RiskRequest` with wire-friendly
    strings instead of SoDEX IntEnums. Decimals come through as JSON
    numbers/strings; Pydantic v2 accepts both transparently.

    Asset is uppercased on the wire (`"BTC"` / `"ETH"`) — risk gate
    enforces the supported set.
    """

    venue: str = Field(..., description="Venue enum value (sodex_spot | sodex_perps)")
    asset: str = Field(..., min_length=1, max_length=10)
    side: ApiSide
    order_type: ApiOrderType
    time_in_force: ApiTimeInForce
    requested_size: Decimal = Field(..., gt=0)
    requested_price: Decimal | None = Field(default=None, gt=0)
    # Perps-only fields. Spot orders MUST leave these unset (or
    # explicitly None); risk gate enforces.
    position_side: ApiPositionSide | None = None
    trigger_type: ApiTriggerType | None = None
    leverage: Decimal | None = Field(default=None, gt=0)
    is_conditional: bool = False
    # Optional linkage to a Signal that drove this order. NULL = ad-hoc.
    signal_id: int | None = Field(default=None, gt=0)
    # PR P1 — stop-loss / take-profit attachment + reduce-only flag.
    # `stop_price` and `stop_type` MUST co-occur (mirrors DB
    # `ck_orders_stop_price_type_consistency`). `parent_order_id`
    # links a child SL/TP to the entry order that opened the position;
    # the risk gate (P1.3) enforces venue + reduce_only consistency.
    stop_price: Decimal | None = Field(default=None, gt=0)
    stop_type: ApiStopType | None = None
    reduce_only: bool = False
    parent_order_id: int | None = Field(default=None, gt=0)

    @model_validator(mode="after")
    def _stop_co_occurrence(self) -> PrepareNewRequest:
        """Mirror `ck_orders_stop_price_type_consistency`: both NULL or
        both set. Catching this at the API boundary turns a 500
        (IntegrityError) into a clean 422 with a useful field path."""
        if (self.stop_price is None) != (self.stop_type is None):
            raise ValueError("stop_price and stop_type must be set together or both be null")
        return self


class SubmitRequest(BaseModel):
    """`POST /api/execution/submit/{order_id}` body.

    Signature shape pinned by regex: `0x01` SoDEX type byte + 65-byte
    ECDSA hex (130 lowercase hex chars). The frontend MUST do the
    `v -= 27` normalization + `0x01` prepend before posting (see
    CLAUDE.md "SoDEX HTTP adapters" §wire-contract).
    """

    signature: str = Field(..., pattern=r"^0x01[0-9a-f]{130}$")


# ---------------------------------------------------------------------------
# Response models
# ---------------------------------------------------------------------------


class PrepareNewResponse(BaseModel):
    """`POST /api/execution/prepare` 200 body.

    Successful path only — risk DENY returns 403 with the reason in
    `detail`; the success body carries the typed-data the wallet
    must sign next.
    """

    order_id: int
    client_order_id: str
    nonce: int
    typed_data: dict


class PrepareCancelResponse(BaseModel):
    """`POST /api/execution/prepare-cancel/{order_id}` 200 body.

    Three shapes:
      - `local_only=True` — PENDING order cancelled in-place; no wallet
        action needed. `typed_data` is None.
      - `replayed=True` — order already terminal (CANCELLED/FILLED/etc).
        No-op. `typed_data` is None.
      - default — `typed_data` carries the cancel typed-data to sign.
    """

    order_id: int
    typed_data: dict | None = None
    client_order_id: str | None = None
    nonce: int | None = None
    local_only: bool = False
    replayed: bool = False


class SubmitResponse(BaseModel):
    """`POST /api/execution/submit{,-cancel}/{order_id}` 200 body."""

    order_id: int
    status: str
    exchange_order_id: str | None = None
    error_message: str | None = None
    replayed: bool = False


# ---------------------------------------------------------------------------
# Read models
# ---------------------------------------------------------------------------


class OrderOut(BaseModel):
    """Single Order row, serialized for `/orders` + `/orders/{id}`.

    Mirrors `models.Order` columns the FE cares about; omits internal
    EIP-712 payload (TEXT, not user-facing) + raw request/response
    (operator-debug surface, exposed via admin only).
    """

    id: int
    user_id: int | None
    signal_id: int | None
    venue: str
    asset: str
    side: str
    order_type: str
    time_in_force: str
    requested_size: Decimal
    requested_price: Decimal | None
    filled_size: Decimal | None
    filled_price: Decimal | None
    filled_value: Decimal | None
    fees: Decimal | None
    status: str
    exchange_order_id: str | None
    client_order_id: str
    error_message: str | None
    paper_trade: bool
    account_id: int | None
    symbol_id: int | None
    nonce: int | None
    nonce_expires_at: datetime | None
    expires_at: datetime | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class PaginatedOrders(BaseModel):
    """`GET /api/execution/orders` paginated response."""

    items: list[OrderOut]
    total: int
    limit: int
    offset: int


class PositionOut(BaseModel):
    """Single Position row."""

    id: int
    user_id: int | None
    venue: str
    asset: str
    side: str
    size: Decimal
    entry_price: Decimal
    stop_loss: Decimal | None
    take_profit: Decimal | None
    status: str
    paper_trade: bool
    leverage: Decimal | None
    opened_at: datetime
    closed_at: datetime | None
    close_price: Decimal | None
    realized_pnl: Decimal | None

    model_config = {"from_attributes": True}


class PositionsResponse(BaseModel):
    """`GET /api/execution/positions` — open positions only."""

    items: list[PositionOut]


class SymbolOut(BaseModel):
    """`GET /api/execution/symbols` row.

    Drives the FE dropdown — venue/asset/name/symbol_id is all the FE
    needs to build an order against the right market. Internal `raw`
    payload from the cache is intentionally not exposed.
    """

    venue: str
    symbol_id: int
    name: str
    asset: str


class SymbolsResponse(BaseModel):
    items: list[SymbolOut]


# ---------------------------------------------------------------------------
# Venue value list (for runtime validation in routes — schema doesn't
# pin venue to an enum because the wire string is `Venue.value`, not
# the Venue enum name).
# ---------------------------------------------------------------------------


VALID_VENUES: frozenset[str] = frozenset(item.value for item in Venue)


# ---------------------------------------------------------------------------
# Limits + usage (P0) — GET /api/execution/limits
# ---------------------------------------------------------------------------


class ExecutionLimitsResponse(BaseModel):
    """Risk caps + the user's current usage against them.

    All notional figures are gross, both-sides, leverage-excluded over the
    rolling 24h window (the exact basis the risk gate enforces — see
    `pipeline.execution.risk.compute_usage`). `per_symbol_*` is populated only
    when an `asset` was queried. Decimals serialize as JSON strings (same
    convention as `OrderOut`)."""

    max_open_orders: int
    open_orders_used: int
    daily_notional_cap: Decimal
    daily_notional_used: Decimal
    per_symbol_cap: Decimal
    per_symbol_used: Decimal | None
    asset: str | None
    max_leverage: int


# ---------------------------------------------------------------------------
# Account summary (P1/P2) — GET /api/execution/account-summary
# ---------------------------------------------------------------------------


class BalanceOut(BaseModel):
    """One spot balance row. `available = total - locked`."""

    asset: str
    total: Decimal
    locked: Decimal
    available: Decimal


class FeeOut(BaseModel):
    """Maker/taker fee rates (fractions, e.g. 0.0005 = 5 bps)."""

    maker_rate: Decimal
    taker_rate: Decimal


class MarkPriceOut(BaseModel):
    """Perps mark price + funding for one symbol. `asset` is the canonical
    base (e.g. "BTC") extracted from `symbol` for client-side joins."""

    symbol: str
    asset: str
    mark_price: Decimal
    funding_rate: Decimal
    next_funding_time: int


class AccountSummaryResponse(BaseModel):
    """Aggregated SoDEX read state for the Execute page: spot balances, fee
    tier, and perps mark prices (for live uPnL + funding). Any field can be
    empty/None independently — a venue with no account 404s to empty rather
    than failing the whole summary."""

    spot_balances: list[BalanceOut]
    fee: FeeOut | None
    mark_prices: list[MarkPriceOut]

File: frontend/src/api/execution.ts
/**
 * Typed wrappers for the PR D.4 execution + wallet routes.
 *
 * Types mirror `etfpulse/api/schemas/execution.py` and
 * `etfpulse/api/schemas/wallet.py` — keep in sync when the backend
 * shapes change. Anti-drift: a backend response_model change MUST
 * land here in the same PR.
 *
 * Wire-format reminders:
 *   - All decimals come through as strings (Pydantic serialises Decimal
 *     as string to avoid float precision loss). The FE treats them as
 *     strings — number parsing happens at display time, not on the
 *     query path, so the user sees the exact backend value.
 *   - `venue` is the StrEnum value ("sodex_spot" | "sodex_perps").
 *   - `signature` posted to /submit MUST be the 0x01-prefixed +
 *     v-normalized form from `toSodexTypedSignature`.
 */

import { apiGet, apiPost } from './client';

// ---------------------------------------------------------------------------
// Wallet (D.4.2)
// ---------------------------------------------------------------------------

export interface WalletMeResponse {
  user_id: number;
  wallet_address: string | null;
  sodex_account_id: number | null;
  paper_trade: boolean;
  sodex_spot_api_key_name: string | null;
  sodex_perps_api_key_name: string | null;
}

export interface SetApiKeyRequest {
  venue: 'sodex_spot' | 'sodex_perps';
  api_key_name: string;
  sodex_account_id: number;
}

export interface SetApiKeyResponse {
  ok: true;
  venue: string;
  api_key_name: string;
  sodex_account_id: number;
}

/**
 * One named API key as returned by `/api/wallet/sodex-bootstrap`.
 * Mirrors `etfpulse/api/schemas/wallet.py:APIKeyOut` field-for-field.
 * `expires_at: 0` means "never expires" per SoDEX api.md.
 */
export interface APIKeyOut {
  name: string;
  public_key: string;
  expires_at: number;
}

/**
 * Response shape of `GET /api/wallet/sodex-bootstrap`. Mirrors
 * `SodexBootstrapResponse` on the backend. Any of the three fields
 * can independently land in its "missing" state:
 *   - account_id null: wallet has no SoDEX account yet
 *   - spot_keys / perps_keys empty: no named keys on that venue
 */
export interface SodexBootstrapResponse {
  account_id: number | null;
  spot_keys: APIKeyOut[];
  perps_keys: APIKeyOut[];
}

// ---------------------------------------------------------------------------
// Execution (D.4.3)
// ---------------------------------------------------------------------------

export type Venue = 'sodex_spot' | 'sodex_perps';
export type Side = 'buy' | 'sell';
export type OrderType = 'limit' | 'market';
export type TimeInForce = 'gtc' | 'ioc' | 'fok' | 'gtx';
export type PositionSide = 'both' | 'long' | 'short';

/** PR P1 — stop-attachment kinds. Mirrors the backend `StopType`
 *  StrEnum (`'stop_loss'` / `'take_profit'`). Mirror invariant is
 *  cross-language; the BE-side `ApiStopType` test pins it. */
export type StopType = 'stop_loss' | 'take_profit';

/** Price feed that arms a perps stop. Mirrors backend `ApiTriggerType`.
 *  Only `mark_price` is supported for placement today (the risk gate
 *  rejects the other two). */
export type TriggerType = 'mark_price' | 'last_price' | 'index_price';

export interface PrepareNewRequest {
  venue: Venue;
  asset: string;
  side: Side;
  order_type: OrderType;
  time_in_force: TimeInForce;
  requested_size: string;
  requested_price?: string | null;
  position_side?: PositionSide | null;
  trigger_type?: TriggerType | null;
  is_conditional?: boolean;
  leverage?: string | null;
  signal_id?: number | null;
  // PR P1 — stop-attachment + reduce-only + child linkage.
  // `stop_price` and `stop_type` MUST co-occur (mirrors BE
  // `ck_orders_stop_price_type_consistency`). A stop order MUST also
  // carry `trigger_type` + `is_conditional` (PR P1-fix.CRIT-1) — the
  // gateway needs the trigger feed and the signed payload carries the
  // stop fields. `parent_order_id` links a child SL/TP to the entry
  // order that opened the position. All perps-only in V1; spot rejects
  // them at the risk gate.
  stop_price?: string | null;
  stop_type?: StopType | null;
  reduce_only?: boolean;
  parent_order_id?: number | null;
}

// EIP-712 typed-data envelope viem's `signTypedData` consumes. The
// backend hands this back verbatim from `pipeline.execution.builders`.
export interface TypedData {
  types: Record<string, { name: string; type: string }[]>;
  primaryType: string;
  domain: {
    name: string;
    version: string;
    chainId: number;
    verifyingContract: string;
  };
  message: Record<string, unknown>;
}

export interface PrepareNewResponse {
  order_id: number;
  client_order_id: string;
  nonce: number;
  typed_data: TypedData;
}

export interface PrepareCancelResponse {
  order_id: number;
  typed_data: TypedData | null;
  client_order_id: string | null;
  nonce: number | null;
  local_only: boolean;
  replayed: boolean;
}

export interface SubmitResponse {
  order_id: number;
  status: string;
  exchange_order_id: string | null;
  error_message: string | null;
  replayed: boolean;
}

export interface OrderOut {
  id: number;
  user_id: number | null;
  signal_id: number | null;
  venue: Venue;
  asset: string;
  side: string;
  order_type: string;
  time_in_force: string;
  requested_size: string;
  requested_price: string | null;
  filled_size: string | null;
  filled_price: string | null;
  filled_value: string | null;
  fees: string | null;
  status: string;
  exchange_order_id: string | null;
  client_order_id: string;
  error_message: string | null;
  paper_trade: boolean;
  account_id: number | null;
  symbol_id: number | null;
  nonce: number | null;
  nonce_expires_at: string | null;
  expires_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface PaginatedOrders {
  items: OrderOut[];
  total: number;
  limit: number;
  offset: number;
}

export interface PositionOut {
  id: number;
  user_id: number | null;
  venue: Venue;
  asset: string;
  side: string;
  size: string;
  entry_price: string;
  stop_loss: string | null;
  take_profit: string | null;
  status: string;
  paper_trade: boolean;
  leverage: string | null;
  opened_at: string;
  closed_at: string | null;
  close_price: string | null;
  realized_pnl: string | null;
}

export interface PositionsResponse {
  items: PositionOut[];
}

export interface SymbolOut {
  venue: Venue;
  symbol_id: number;
  name: string;
  asset: string;
}

export interface SymbolsResponse {
  items: SymbolOut[];
}

// ---------------------------------------------------------------------------
// Wallet API wrappers
// ---------------------------------------------------------------------------

export function fetchWalletMe(): Promise<WalletMeResponse> {
  return apiGet<WalletMeResponse>('/api/wallet/me');
}

export function postWalletApiKey(req: SetApiKeyRequest): Promise<SetApiKeyResponse> {
  return apiPost<SetApiKeyResponse>('/api/wallet/api-key', req);
}

export function fetchSodexBootstrap(): Promise<SodexBootstrapResponse> {
  return apiGet<SodexBootstrapResponse>('/api/wallet/sodex-bootstrap');
}

export interface RequestLiveRequest {
  note?: string;
}

export interface RequestLiveResponse {
  ok: true;
  message: string;
}

export function postRequestLive(req: RequestLiveRequest): Promise<RequestLiveResponse> {
  return apiPost<RequestLiveResponse>('/api/wallet/request-live', req);
}

// ---------------------------------------------------------------------------
// Execution API wrappers
// ---------------------------------------------------------------------------

export function postPrepareNew(req: PrepareNewRequest): Promise<PrepareNewResponse> {
  return apiPost<PrepareNewResponse>('/api/execution/prepare', req);
}

export function postSubmitNew(orderId: number, signature: string): Promise<SubmitResponse> {
  return apiPost<SubmitResponse>(`/api/execution/submit/${orderId}`, { signature });
}

export function postPrepareCancel(orderId: number): Promise<PrepareCancelResponse> {
  return apiPost<PrepareCancelResponse>(`/api/execution/prepare-cancel/${orderId}`);
}

export function postSubmitCancel(orderId: number, signature: string): Promise<SubmitResponse> {
  return apiPost<SubmitResponse>(`/api/execution/submit-cancel/${orderId}`, { signature });
}

/** PR P1.4 — close-position. Returns the same PrepareNew shape as
 *  /prepare; the wallet still has to sign + submit via /submit. */
export function postClosePosition(positionId: number): Promise<PrepareNewResponse> {
  return apiPost<PrepareNewResponse>(`/api/execution/close-position/${positionId}`);
}

export function fetchOrders(params?: {
  venue?: Venue;
  status?: string;
  limit?: number;
  offset?: number;
}): Promise<PaginatedOrders> {
  return apiGet<PaginatedOrders>('/api/execution/orders', params);
}

export function fetchPositions(): Promise<PositionsResponse> {
  return apiGet<PositionsResponse>('/api/execution/positions');
}

export function fetchSymbols(venue?: Venue): Promise<SymbolsResponse> {
  return apiGet<SymbolsResponse>('/api/execution/symbols', venue ? { venue } : undefined);
}
